File: app/src/test/java/com/maximillionsnyder/umafinidad/AptitudesTest.kt
package com.maximillionsnyder.umafinidad

import com.maximillionsnyder.umafinidad.data.CharacterDto
import com.maximillionsnyder.umafinidad.data.MemberDto
import com.maximillionsnyder.umafinidad.data.RelationDto
import com.maximillionsnyder.umafinidad.data.deserializarAptitudes
import com.maximillionsnyder.umafinidad.data.jsonParser
import com.maximillionsnyder.umafinidad.data.toDomain
import com.maximillionsnyder.umafinidad.domain.AffinityModel
import com.maximillionsnyder.umafinidad.domain.aptitudValida
import com.maximillionsnyder.umafinidad.domain.aptitudesDestacadas
import kotlinx.serialization.decodeFromString
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test

/* Tests de aptitudes (pista / distancia / estilo de la carta base):
   validación del JSON embebido, accessor del modelo y destacadas. */
class AptitudesTest {

    private lateinit var modelo: AffinityModel
    private lateinit var aptitudes: Map<Int, List<String>>

    private fun recurso(ruta: String): String =
        javaClass.getResourceAsStream("/$ruta")!!.bufferedReader().use { it.readText() }

    @Before
    fun preparar() {
        val characters = jsonParser.decodeFromString<List<CharacterDto>>(recurso("data/characters.json"))
        val relations = jsonParser.decodeFromString<List<RelationDto>>(recurso("data/succession_relation.json"))
        val members = jsonParser.decodeFromString<List<MemberDto>>(recurso("data/succession_relation_member.json"))
        modelo = AffinityModel(
            characters.map { it.toDomain() },
            relations.map { it.toDomain() },
            members.map { it.toDomain() },
        )
        aptitudes = deserializarAptitudes(recurso("data/aptitudes.json"))
    }

    /* ===== Deserialización ===== */

    @Test
    fun clavesConvertidasAInt() {
        assertTrue(aptitudes.isNotEmpty())
        assertTrue(1001 in aptitudes)
        assertTrue(aptitudes.keys.all { it > 0 })
    }

    @Test
    fun todasLasEntradasSonValidas() {
        assertTrue(aptitudes.values.all { aptitudValida(it) })
    }

    /* La tabla cubre a los personajes jugables (tolerancia mínima por
       desfasajes de hasta un día entre los JSON de GameTora). */
    @Test
    fun coberturaDeJugables() {
        val jugables = modelo.personajes.filter { it.playable == true && it.active == true }.map { it.charId }.toSet()
        val extras = aptitudes.keys - jugables
        assertTrue("demasiadas entradas de no jugables: $extras", extras.size <= 5)
        assertTrue(
            "cobertura insuficiente: ${jugables.size - aptitudes.size} jugables sin aptitudes",
            jugables.size - aptitudes.size <= 3,
        )
    }

    /* ===== Accessor del modelo ===== */

    @Test
    fun specialWeekVerificadoContraGameTora() {
        val m = AffinityModel(
            modelo.personajes,
            emptyList(),
            emptyList(),
            aptitudes,
        )
        assertEquals(
            listOf("A", "G", "F", "C", "A", "A", "G", "A", "A", "C"),
            m.aptitudesDe(1001),
        )
    }

    @Test
    fun idDesconocidoOSinTablaDevuelveNull() {
        val m = AffinityModel(modelo.personajes, emptyList(), emptyList(), aptitudes)
        assertNull(m.aptitudesDe(999999999))
        /* Sin la tabla (constructor por defecto) nunca crashea. */
        assertNull(modelo.aptitudesDe(1001))
    }

    /* ===== Destacadas (A o B) ===== */

    @Test
    fun destacadasDeSpecialWeek() {
        assertEquals(
            listOf(0, 4, 5, 7, 8), /* turf, media, larga, vanguardia, remate */
            aptitudesDestacadas(listOf("A", "G", "F", "C", "A", "A", "G", "A", "A", "C")),
        )
    }

    @Test
    fun destacadasIncluyenLaLetraB() {
        assertEquals(
            listOf(1, 2),
            aptitudesDestacadas(listOf("C", "B", "A", "C", "C", "C", "C", "C", "C", "C")),
        )
    }

    @Test
    fun destacadasVaciasCuandoNoDestacaEnNada() {
        assertTrue(aptitudesDestacadas(List(10) { "C" }).isEmpty())
    }

    /* ===== Validación ===== */

    @Test
    fun aptitudInvalidaRechazada() {
        assertTrue(aptitudValida(List(10) { "A" }))
        assertTrue(!aptitudValida(List(9) { "A" }))
        assertTrue(!aptitudValida(List(10) { "S" }))
        assertTrue(!aptitudValida(listOf("A", "G", "F", "C", "A", "A", "G", "A", "A", "AA")))
    }

    /* Especialistas conocidos: la dirt de Smart Falcon (1046) es A y el
       estilo de Twin Turbo (1066) es 逃げ puro. */
    @Test
    fun especialistasConocidos() {
        val smartFalcon = aptitudes[1046]
        assertNotNull(smartFalcon)
        assertEquals("A", smartFalcon!![1])

        val twinTurbo = aptitudes[1066]
        assertNotNull(twinTurbo)
        assertEquals("A", twinTurbo!![6])
        assertEquals("G", twinTurbo[7])
        assertEquals("G", twinTurbo[8])
        assertEquals("G", twinTurbo[9])
    }
}

File: sitio/src/lib/domain/__tests__/aptitudes.test.ts
/* Aptitudes (pista / distancia / estilo de la carta base).
   Espejo de AptitudesTest.kt. */

import { describe, expect, it } from 'vitest'
import { aptitudValida, aptitudesDestacadas } from '../aptitudes'
import { AffinityModel } from '../affinity'
import { aptitudes, characters, ids, modelo, relations, members } from '../../../test/data'

describe('deserialización de aptitudes', () => {
  it('las claves se convierten a números', () => {
    expect(aptitudes.size).toBeGreaterThan(0)
    expect(aptitudes.has(1001)).toBe(true)
    expect([...aptitudes.keys()].every((k) => k > 0)).toBe(true)
  })

  it('todas las entradas son válidas', () => {
    expect([...aptitudes.values()].every((apt) => aptitudValida(apt))).toBe(true)
  })

  it('la tabla cubre a los jugables', () => {
    const jugables = new Set(ids)
    const extras = [...aptitudes.keys()].filter((id) => !jugables.has(id))
    expect(extras.length, `entradas de no jugables: ${extras}`).toBeLessThanOrEqual(5)
    expect(jugables.size - aptitudes.size).toBeLessThanOrEqual(3)
  })
})

describe('accessor del modelo', () => {
  it('Special Week verificado contra GameTora', () => {
    const m = new AffinityModel(characters, [], [], aptitudes)
    expect(m.aptitudesDe(1001)).toEqual(['A', 'G', 'F', 'C', 'A', 'A', 'G', 'A', 'A', 'C'])
  })

  it('id desconocido o sin tabla devuelve null', () => {
    const m = new AffinityModel(characters, [], [], aptitudes)
    expect(m.aptitudesDe(999999999)).toBeNull()
    /* Sin la tabla (constructor por defecto) nunca crashea. */
    expect(modelo().aptitudesDe(1001)).toBeNull()
  })

  it('el modelo real (con relaciones) también expone aptitudes', () => {
    const m = new AffinityModel(characters, relations, members, aptitudes)
    expect(m.aptitudesDe(1001)).not.toBeNull()
  })
})

describe('aptitudes destacadas (A o B)', () => {
  it('destacadas de Special Week', () => {
    expect(aptitudesDestacadas(['A', 'G', 'F', 'C', 'A', 'A', 'G', 'A', 'A', 'C'])).toEqual([
      0, 4, 5, 7, 8, /* turf, media, larga, vanguardia, remate */
    ])
  })

  it('incluyen la letra B', () => {
    expect(aptitudesDestacadas(['C', 'B', 'A', 'C', 'C', 'C', 'C', 'C', 'C', 'C'])).toEqual([1, 2])
  })

  it('vacías cuando no destaca en nada', () => {
    expect(aptitudesDestacadas(Array(10).fill('C'))).toEqual([])
  })
})

describe('validación', () => {
  it('rechaza aptitudes inválidas', () => {
    expect(aptitudValida(Array(10).fill('A'))).toBe(true)
    expect(aptitudValida(Array(9).fill('A'))).toBe(false)
    expect(aptitudValida(Array(10).fill('S'))).toBe(false)
    expect(aptitudValida(['A', 'G', 'F', 'C', 'A', 'A', 'G', 'A', 'A', 'AA'])).toBe(false)
  })

  it('especialistas conocidos', () => {
    const smartFalcon = aptitudes.get(1046)
    expect(smartFalcon).toBeDefined()
    expect(smartFalcon![1]).toBe('A')

    const twinTurbo = aptitudes.get(1066)
    expect(twinTurbo).toBeDefined()
    expect(twinTurbo![6]).toBe('A')
    expect(twinTurbo![7]).toBe('G')
    expect(twinTurbo![8]).toBe('G')
    expect(twinTurbo![9]).toBe('G')
  })
})
